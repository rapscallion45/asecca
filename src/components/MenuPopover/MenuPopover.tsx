import { FC, ReactNode } from 'react';
import { Popover } from '@mui/material';
import { alpha, styled } from '@mui/material/styles';

/**
 * Arrow Style
 *
 * Custom menu popover anchor arrow styling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.2
 *
 * @component
 * @returns {Component} - styled anchor arrow component
 */
const ArrowStyle = styled('span')(({ theme }) => ({
  [theme.breakpoints.up('sm')]: {
    top: -7,
    zIndex: 1,
    width: 12,
    right: 20,
    height: 12,
    content: "''",
    position: 'absolute',
    borderRadius: '0 0 4px 0',
    transform: 'rotate(-135deg)',
    background: theme.palette.background.paper,
    borderRight: `solid 1px ${alpha(theme.palette.grey[500], 0.12)}`,
    borderBottom: `solid 1px ${alpha(theme.palette.grey[500], 0.12)}`,
  },
}));

/**
 * Menu Popover Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.2
 *
 * @typedef IMenuPopoverProps
 * @prop {boolean} open - popover open flag
 * @prop {ReactNode} children - child nodes
 * @prop {any} sx - custom styling options
 * @prop {any} other - passed props
 * @prop {any} onClose - popover close callback
 * @prop {any} anchorE1 - popover anchor element
 * @prop {any} x - all other props
 */
interface IMenuPopoverProps {
  open: boolean;
  children?: ReactNode;
  sx?: any;
  other?: any;
  onClose?: () => void;
  anchorEl: any;
  // all other props
  [x: string]: any;
}

/**
 * Menu Popover
 *
 * Global application menu popover component
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @component
 * @param {IMenuPopoverProps} props - component props
 * @returns {FC} - menu popover component
 */
const MenuPopover: FC<IMenuPopoverProps> = (props) => {
  const { open, children = null, sx = null, ...other } = props;

  return (
    <Popover
      open={open}
      anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      PaperProps={{
        sx: {
          mt: 1.5,
          pt: 1,
          ml: 0.5,
          overflow: 'inherit',
          border: 'none',
          width: 200,
          ...sx,
        },
      }}
      {...other}
    >
      <ArrowStyle className="arrow" />
      {children}
    </Popover>
  );
};

export default MenuPopover;
