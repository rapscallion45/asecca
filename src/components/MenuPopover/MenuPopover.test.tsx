import React from 'react';
import { render } from '@testing-library/react';
import MenuPopover from './MenuPopover';

/**
 * Menu Popover Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Menu Popover', () => {
  it('Renders correctly', async () => {
    /** perform snapshot test */
    const tree = render(<MenuPopover open anchorEl />);
    expect(tree).toMatchSnapshot();
  });
});
