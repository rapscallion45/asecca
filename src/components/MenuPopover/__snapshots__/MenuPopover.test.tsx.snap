// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Popover Renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiPopover-root MuiModal-root css-1i0e6f3-MuiModal-root-MuiPopover-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-1i42k3j-MuiPaper-root-MuiPopover-paper"
        style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
        tabindex="-1"
      >
        <span
          class="arrow css-1xmfx9g"
        />
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
